lib/ffi-icu/time_formatting.rb

```diff
@@ -215,6 +229,109 @@ def skeleton_format(skeleton_pattern_str, locale)
           retry
         end
       end
+
+      private
+
+      # Converts the current pattern to a pattern that takes the desired hour cycle
+      # into account. This is needed because most of the standard patterns in ICU
+      # contain either h (12 hour) or H (23 hour) in them, instead of j (locale-
+      # specified hour cycle). This means if you use a locale with an @hours=h12
+      # keyword in it, for example, it would normally be totally ignored by ICU.
+      #
+      # This is the same fixup done by Firefox:
+      # https://github.com/tc39/ecma402/issues/665#issuecomment-1084833809
+      # https://searchfox.org/mozilla-central/rev/625c3d0c8ae46502aed83f33bd530cb93e926e9f/intl/components/src/DateTimeFormat.cpp#282-323
+      def replace_hour_symbol!
+        # Short circuit this case - nil means "use whatever is in the pattern already", so
+        # no need to actually run any of this implementation.
+        return unless @hour_cycle
+
+        # Get the current pattern and convert to a skeleton
+        skeleton_str = pattern_to_skeleton_uchar(current_pattern_as_uchar).string
+
+        # Manipulate the skeleton to make it work with the correct hour cycle.
+        skeleton_str.gsub!(/[hHkKjJ]/, HOUR_CYCLE_SYMS[@hour_cycle])
+
+        # Either ensure the skeleton has, or does not have, am/pm, as appropriate
+        if ['h11', 'h12'].include?(@hour_cycle)
+          skeleton_str << 'a' unless skeleton_str.include? 'a'
+        else
+          skeleton_str.gsub!('a', '')
+        end
+
+        # Convert the skeleton back to a pattern
+        new_pattern_str = skeleton_to_pattern_uchar(UCharPointer.from_string(skeleton_str)).string
+
+        # We also need to manipulate the _pattern_, a little bit, because (according to Firefox source):
+        #
+        #     Input skeletons don't differentiate between "K" and "h" resp. "k" and "H".
+        #
+        # https://searchfox.org/mozilla-central/rev/625c3d0c8ae46502aed83f33bd530cb93e926e9f/intl/components/src/DateTimeFormat.cpp#183
+        # So, if we put a skeleton with a k in it into getBestPattern, it comes out with a H (and a
+        # skeleton with a K in it comes out with a h). Need to fix this in the generated pattern.
+        if HOUR_CYCLE_SYMS.keys.include?(real_hour_cycle)
+          new_pattern_str.gsub!(/[hHkK]/, HOUR_CYCLE_SYMS[real_hour_cycle])
+        end
```

Comment:
```suggestion
        @hour_cycle = Locale.new(@locale).keyword('hours') if @hour_cycle == :locale

        if HOUR_CYCLE_SYMS.keys.include?(@hour_cycle)
          new_pattern_str.gsub!(/[hHkK]/, HOUR_CYCLE_SYMS[@hour_cycle])
        end
```

Extracting the hour from the locale is simple enough to do inline and we can remove the `real_hour_cycle` method below.

Reply:
I agree that `real_` isn't a great name for something, but it's not quite correct to clobber `@hour_cycle` like this - that will have the effect of using e.g. `h` or `H` in the skeleton instead of `j` the _next_ time `replace_hour_symbol!` is called (for example, if someone calls `set_date_format`